currency/pair.go

```diff
@@ -44,161 +52,45 @@ func NewPairWithDelimiter(base, quote, delimiter string) Pair {
 
 // NewPairFromIndex returns a CurrencyPair via a currency string and specific
 // index
-func NewPairFromIndex(currencyPair, index string) (Pair, error) {
+func NewPairFromIndex(currencyPair, index string) (*Pair, error) {
 	i := strings.Index(currencyPair, index)
 	if i == -1 {
-		return Pair{},
+		return nil,
 			fmt.Errorf("index %s not found in currency pair string", index)
 	}
 	if i == 0 {
 		return NewPairFromStrings(currencyPair[0:len(index)],
-				currencyPair[len(index):]),
-			nil
+			currencyPair[len(index):])
 	}
-	return NewPairFromStrings(currencyPair[0:i], currencyPair[i:]), nil
+	return NewPairFromStrings(currencyPair[0:i], currencyPair[i:])
 }
 
 // NewPairFromString converts currency string into a new CurrencyPair
 // with or without delimeter
-func NewPairFromString(currencyPair string) Pair {
-	delimiters := []string{"_", "-", "/", ":"}
-	var delimiter string
-	for _, x := range delimiters {
-		if strings.Contains(currencyPair, x) {
-			delimiter = x
-			return NewPairDelimiter(currencyPair, delimiter)
+func NewPairFromString(currencyPair string) (*Pair, error) {
+	for x := range delimiters {
+		if strings.Contains(currencyPair, delimiters[x]) {
+			return NewPairDelimiter(currencyPair, delimiters[x]), nil
 		}
 	}
+	if len(currencyPair) < 3 {
+		return nil,
+			fmt.Errorf("cannot produce a currency pair from %s string",
+				currencyPair)
+	}
 	return NewPairFromStrings(currencyPair[0:3], currencyPair[3:])
 }
 
 // NewPairFromFormattedPairs matches a supplied currency pair to a list of pairs
 // with a specific format. This is helpful for exchanges which
 // provide currency pairs with no delimiter so we can match it with a list and
 // apply the same format
-func NewPairFromFormattedPairs(currencyPair string, pairs Pairs, pairFmt PairFormat) Pair {
+func NewPairFromFormattedPairs(currencyPair string, pairs Pairs, pairFmt PairFormat) (*Pair, error) {
 	for x := range pairs {
-		if strings.EqualFold(pairs[x].Format(pairFmt.Delimiter,
-			pairFmt.Uppercase).String(), currencyPair) {
-			return pairs[x]
+		fPair := pairs[x].Format(pairFmt.Delimiter, pairFmt.Uppercase)
```

Comment:
So here, we're creating a whole bunch of new currency pair pointers and then not storing them. So they will eventually just be thrown out when the garbage collector comes